File: raven/tui_rpc/methods/dag.py
"""``dag.get`` / ``dag.node`` RPC handlers.

A ``run_subagent_dag`` run publishes ``dag.*`` progress events while it executes,
and nothing replays them. These two read the run dir instead, so a TUI that was
not listening for every frame is not stuck with whatever it happened to catch:

* ``dag.get`` rebuilds (or repairs) a whole graph. An unfinalized run has no
  manifest, so the instance registry is overlaid on top -- see
  :func:`~raven.agent.subagent_dag._resume.read_run_reconciled`, shared with the
  web surface so a resumed graph means the same thing on both.
* ``dag.node`` pulls one node's *rendered* prompt and its output, neither of
  which the graph carries: the manifest inlines only the leaf nodes' text, and
  the tool result the transcript keeps is clamped to 200 chars.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from raven.agent.subagent_dag._errors import DagValidationError
from raven.agent.subagent_dag._reader import DagReadError
from raven.agent.subagent_dag._resume import read_run_reconciled
from raven.tui_rpc.errors import InternalError

if TYPE_CHECKING:
    from raven.tui_rpc.dispatcher import Dispatcher
    from raven.tui_rpc.methods.session import AgentLoopFactory


def _dag_tool(agent_loop_factory: "AgentLoopFactory | None") -> Any:
    """The live DAG tool, or a typed error naming why there is none.

    The tool is only registered when third-party sub-agents are configured, so
    "not configured" is a normal state -- but one the client can do nothing about
    mid-call, and it must not read as an empty run.
    """
    loop = agent_loop_factory() if agent_loop_factory is not None else None
    tools = getattr(loop, "tools", None) if loop is not None else None
    tool = tools.get("run_subagent_dag") if tools is not None else None
    if tool is None:
        raise InternalError("no live run_subagent_dag tool (configure third-party sub-agents first)")
    return tool


async def dag_get(params: dict, *, agent_loop_factory: "AgentLoopFactory | None" = None) -> dict:
    """One run's structure and per-node state, read back off disk."""
    tool = _dag_tool(agent_loop_factory)
    # The reader raises its own ValueError subclass for a malformed id or a run
    # dir that is gone. Untyped, that reaches the client as a -32603 traceback
    # instead of a message it can show.
    try:
        run = await read_run_reconciled(tool, params.get("run_id", ""), params.get("session_key"))
    except (DagReadError, DagValidationError) as exc:
        raise InternalError(str(exc)) from exc
    return {"run": run}


async def dag_node(params: dict, *, agent_loop_factory: "AgentLoopFactory | None" = None) -> dict:
    """One node's rendered prompt and (head of its) output."""
    tool = _dag_tool(agent_loop_factory)
    try:
        node = await tool.read_node(
            params.get("run_id", ""),
            params.get("node", ""),
            max_output_chars=int(params.get("max_output_chars") or 20000),
        )
    except (DagReadError, DagValidationError) as exc:
        raise InternalError(str(exc)) from exc
    return {"node": node}


def register_dag_methods(
    dispatcher: "Dispatcher",
    *,
    agent_loop_factory: "AgentLoopFactory | None" = None,
) -> None:
    """Register ``dag.get`` / ``dag.node`` on a dispatcher instance."""

    async def _get(params: dict) -> dict:
        return await dag_get(params, agent_loop_factory=agent_loop_factory)

    async def _node(params: dict) -> dict:
        return await dag_node(params, agent_loop_factory=agent_loop_factory)

    dispatcher.register("dag.get", _get)
    dispatcher.register("dag.node", _node)

File: ui-tui/src/app/slash/commands/dag.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 EverMind.
// See NOTICES.md.
//
// `/dag` — the user-initiated half of the DAG surface.
//
// The live `dag.*` events cover a run that behaves. Two things they cannot do:
//
//   * repair a graph whose frames were lost. Nothing replays them, so a gateway
//     that restarted mid-run leaves nodes pinned to `running` forever. `dag.get`
//     re-reads the run dir, where the instance registry supplies the progress
//     the events did not deliver.
//   * show a node's rendered prompt or its output. No event carries either: the
//     manifest inlines only the leaf nodes' text, and the tool result kept in
//     the transcript is clamped to 200 chars.
//   * name the nodes. `run_subagent_dag`'s own call label elides past the third
//     id, and a graph row carries its id only in the block a click opens, so on
//     a graph of any size the refresh listing is the only way to learn the id
//     this command takes as its argument.
//
// Both are user-initiated for a reason: the terminal event handlers are
// synchronous (the turn commits its transcript row inside them), so an RPC
// round-trip cannot be awaited there without reordering the event stream.

import type { DagGetResult, DagNodeResult } from '../../../rpc/index.js'
import type { SlashCommand } from '../types.js'

import { dagRunHeadline, formatDagNodeDetail } from '../../../lib/dagStatus.js'
import { turnController } from '../../turnController.js'
import { getTurnState } from '../../turnStore.js'

// The file itself is uncapped and can be megabytes. This is a transcript, not a
// pager, so ask for a slice that stays readable and say when it was cut.
const NODE_OUTPUT_CHARS = 4000

export const dagCommands: SlashCommand[] = [
  {
    help: "refresh this turn's sub-agent DAG graphs, or show one node's prompt/output",
    name: 'dag',
    run: (arg, ctx) => {
      const { gateway, transcript, ui } = ctx
      const runs = getTurnState().dagRuns
      const node = arg.trim()

      if (runs.length === 0) {
        transcript.sys('no DAG run in this turn')

        return
      }

      if (!node) {
        runs.forEach(run => {
          gateway
            .rpc<DagGetResult>('dag.get', { run_id: run.runId, session_key: ui.sid })
            .then(
              ctx.guarded<DagGetResult>(result => {
                turnController.applyDagSnapshot(result.run)
                // Read the tally back off the store rather than the response, so
                // the line cannot disagree with the graph that just re-rendered.
                const refreshed = getTurnState().dagRuns.find(item => item.runId === run.runId)
                transcript.sys(`${run.runId}: ${refreshed ? dagRunHeadline(refreshed) : 'refreshed'}`)

                // The ids, because this command's own argument is one and there
                // is no other keyboard route to them: the tool row's label
                // elides past the third, and a graph row shows its id only once
                // expanded, which takes a mouse.
                if (refreshed && refreshed.nodes.length > 0) {
                  transcript.sys(`  nodes: ${refreshed.nodes.map(item => item.id).join(', ')}`)
                }
              })
            )
            .catch((err: unknown) => {
              // Leave the graph as it was: a run dir that was cleaned up is not
              // a reason to blank what the user can still see.
              if (!ctx.stale()) {
                transcript.sys(`dag refresh failed for ${run.runId}: ${String(err)}`)
              }
            })
        })

        return
      }

      // A node is addressed by id alone, so it has to be found in a graph. The
      // most recent run wins when several are open -- the same one whose live
      // panel is on screen.
      const owner = [...runs].reverse().find(run => run.nodes.some(item => item.id === node))

      if (!owner) {
        transcript.sys(`no node '${node}' in this turn's DAG runs`)

        return
      }

      gateway
        .rpc<DagNodeResult>('dag.node', {
          max_output_chars: NODE_OUTPUT_CHARS,
          node,
          run_id: owner.runId,
          session_key: ui.sid
        })
        .then(
          ctx.guarded<DagNodeResult>(result => {
            transcript.page(formatDagNodeDetail(result.node), `${node} @ ${owner.runId}`)
          })
        )
        .catch((err: unknown) => {
          if (!ctx.stale()) {
            transcript.sys(`dag node read failed for ${node}: ${String(err)}`)
          }
        })
    },
    usage: '/dag [node]'
  }
]
